### PR TITLE
Demo search on frontend

### LINKED ISSUE #2057: Implement demo search on frontend
We need to support an ability to search the demo by organization and project as a part of `DemoMainView`

### PR DESCRIPTION
**Pull request by sanyavertolet**
This PR closes #2057

### What's done:
* Replaced stub `demoList` with functional one
* Added demo filtering
* Implemented `useDebouncedDeferredRequest`

### How it looks:
https://user-images.githubusercontent.com/35039155/228872066-a0409284-e4d6-403e-9c6a-ea9184bc7b6a.mov

### CODE REVIEW

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt [hunk 1]
```diff
@@ -88,9 +91,29 @@ class DemoService(
...
-     * @return list of [Demo]s that are stored in database
+     * @param demoFilter [DemoFilter]
+     * @param pageSize amount of [Demo]s that should be fetched
+     * @return list of [Demo]s that match [DemoFilter]
      */
-    fun getAllDemos(): List<Demo> = demoRepository.findAll()
+    fun getFiltered(demoFilter: DemoFilter, pageSize: Int): Page<Demo> = demoRepository.findAll({ root, _, cb ->
+        with(demoFilter) {
+            val organizationNamePredicate = if (organizationName.isBlank()) {
+                cb.and()
```
Comment (nulls) [inline save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:101]: what's about status?

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt [hunk 2]
```diff
@@ -131,7 +154,6 @@ class DemoService(
      * @param projectName saveourtool project name
      * @return [Demo] connected with project [organizationName]/[projectName] or null if not present
      */
-    @Transactional
```
Comment (nulls) [inline save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:134]: or `@Transactional(readonly=true)`?

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/controller/DemoController.kt [hunk 3]
```diff
@@ -36,13 +37,19 @@ class DemoController(
...
+     * @param demoAmount number of [DemoDto]s that should be fetched, [DemoDto.DEFAULT_FETCH_NUMBER] by default
+     * @return all [DemoDto]s matching [filter] as [Flux]
      */
-    @GetMapping("/active")
-    fun active(): Flux<DemoDto> = blockingToFlux {
-        demoService.getAllDemos().map { it to demoService.getStatus(it).block() }
+    @PostMapping("/demo-list")
+    fun getFilteredDemoList(
+        @RequestBody(required = false) filter: DemoFilter = DemoFilter.any,
+        @RequestParam(required = false, defaultValue = DemoDto.DEFAULT_FETCH_NUMBER.toString())
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/controller/DemoController.kt:47]: It seems that the [issue](https://github.com/spring-projects/spring-framework/issues/21139) is still opened so will make `filter` nullable and perform null-check

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:101]: Right now I check statuses after the filtered demo is fetched as I do not store the status in database due to no sync solution (right now there is no way to keep data up to date, need to implement some kind of service that does this task repeatedly over time). If you find it better, I could move status filtering to this method, but the point is that status check right now is an expensive operation

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:134]: Probably 😃

Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:134]: The point is that Idea gives me an error:
```
@Transactional self-invocation (in effect, a method within the target object calling another method of the target object) does not lead to an actual transaction at runtime 
```
 on `findBySaveourtoolProject` invocation in method:
```Kotlin
fun findBySaveourtoolProjectOrNotFound(
    organizationName: String,
    projectName: String,
    lazyMessage: () -> String,
): Mono<Demo> = blockingToMono {
    findBySaveourtoolProject(organizationName, projectName)
}.switchIfEmptyToNotFound(lazyMessage)
```

Comment (Cheshiriks) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:134]: You cannot call a transactional method from another method in the same class

Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:134]: Discussed offline - decided to ignore the warning 😁

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/service/DemoService.kt:101]: Will fix that in the next PR.

File: save-demo/src/main/kotlin/com/saveourtool/save/demo/controller/DemoController.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-demo/src/main/kotlin/com/saveourtool/save/demo/controller/DemoController.kt:48]: Done

**Review by nulls — APPROVED** [review]

**Pull request merged**